executor/prebackup.go

```diff
@@ -0,0 +1,142 @@
+
+	return b.startAllPreBackupDeployments(deployments)
+}
+
+// startAllPreBackupDeployments attempts to start all given deployment templates.
+// If some of them are already existing, their status will determine whether the backup is considered failed.
+// It returns true if all deployments are ready, also the ConditionPreBackupPodsReady will be set to ReasonReady.
+// Otherwise it returns false, indicating that not all PreBackup deployment(s) are ready.
+func (b *BackupExecutor) startAllPreBackupDeployments(deployments []appsv1.Deployment) (bool, error) {
+	existingDeployments := make([]*appsv1.Deployment, 0)
+	for _, deployment := range deployments {
+		exists, err := b.createPreBackupDeployment(deployment)
+		if err != nil {
+			return false, err
+		}
+		if exists {
+			deployment, err := b.fetchPreBackupDeployment(deployment.Name, deployment.Namespace)
+			if err != nil {
+				return false, err
+			}
+			existingDeployments = append(existingDeployments, deployment)
+		} else {
+			b.Log.Info("started pre backup pod", "namespace", deployment.Namespace, "name", deployment.Name)
+		}
+	}
+	if ready, err := b.allDeploymentsAreReady(existingDeployments); ready {
+		b.Log.Info("pre backup pod(s) now ready")
+		b.SetConditionTrue(k8upv1alpha1.ConditionPreBackupPodsReady, k8upv1alpha1.ReasonReady)
+		return true, nil
+	} else if err == nil {
+		b.SetConditionUnknownWithMessage(k8upv1alpha1.ConditionPreBackupPodsReady, k8upv1alpha1.ReasonWaiting, "waiting for %d PreBackupPods to become ready", len(deployments))
+	}
+	return false, nil
+}
+
+// allDeploymentsAreReady returns true if all given pre backup deployments have replicas in ready state.
+// If one of the deployment has failed (determined via "Progressing" condition), then it returns false with an error.
+// In that case, a backup is considered failed completely and status conditions will be set accordingly.
+// If one deployment is neither failed nor ready, then it returns false without error, indicating that it's still waiting.
+func (b *BackupExecutor) allDeploymentsAreReady(deployments []*appsv1.Deployment) (bool, error) {
+	readyCounter := 0
+	for _, deployment := range deployments {
+		ready, err := isPreBackupDeploymentReady(deployment)
+		log := b.Log.WithValues("preBackup", deployment.Name+"/"+deployment.Namespace)
+		if err != nil {
+			log.Info("backup failed: deadline exceeded on pre backup deployment")
+			b.SetConditionFalseWithMessage(k8upv1alpha1.ConditionPreBackupPodsReady, k8upv1alpha1.ReasonFailed, err.Error())
+			b.SetConditionTrueWithMessage(k8upv1alpha1.ConditionReady, k8upv1alpha1.ReasonFailed, err.Error())
+			b.deletePreBackupDeployment(deployment)
+			return false, err
+		}
+		if ready {
+			log.V(2).Info("pre backup pod is in ready state")
+			readyCounter++
+		} else {
+			log.Info("still waiting on pre backup pod...")
+			return false, nil
+		}
+	}
+	return readyCounter == len(deployments) && readyCounter > 0, nil
```

Comment:
I would invert this: Instead of a `readyCounter` you could use a `readyCountdown`:

Initialize the countdown with `len(deployments)` and subtract 1 for every deployment that is ready.
If – in the end – the countdown is 0, then all deployments are ready.

Reply:
After looking at it again, I think it's redundant anyway. Because every non-ready condition leads to an abort of the `for` loop. So if I see this correctly, neither a countdown nor a countup is required and this can be shortened to:

```suggestion
	return true, nil
```

Reply:
you might be right, I'll check